onnxruntime/core/optimizer/compute_optimizer/shared_utils.cc

```diff
@@ -99,6 +99,54 @@ Node* InsertIntermediateNodeOnDestInput(Graph& graph,
   return &new_node;
 }
 
+std::pair<bool, std::vector<DimCompareRet>> CompareInputShapeWithOutputShape(
+    const ONNX_NAMESPACE::TensorShapeProto* full_broadcasted_shape,
+    const ONNX_NAMESPACE::TensorShapeProto* target_shape) {
+  int full_rank = full_broadcasted_shape->dim_size();
+  int target_rank = target_shape->dim_size();
+
+  if (target_rank > full_rank) {
+    return std::make_pair<bool, std::vector<DimCompareRet>>(false, {});
+  }
+
+  std::vector<DimCompareRet> rets(full_rank);
+  for (int i = -1; i >= -full_rank; --i) {
```

Comment:
What is the motivation behind using the negatives?

Reply:
For two broadcasting shapes, full broadcasted shape v.s. target shape. They are aligned from the right side, so we check the equality from the right side  full broadcasted shape[-1] ==  target shape[-1]. Let me add a comment for this.